### PR TITLE
closest pair of points algo

### PR DESCRIPTION
**Pull request by Dharni0607**
added closest pair of points algorithm under divide and conquer.
Please review.
;)

### CODE REVIEW

File: divide_and_conquer/closest_pair_of_points.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+The algorithm finds distance between closest pair of points in the given n points.
+Approach used -> Divide and conquer 
+The points are sorted based on x-cords 
+& by applying divide and conquer approach, 
+minimum distance is obtained recursively.
+
+Edge case: closest points lie on different sides of partition
+This case handled by forming a strip of points 
+which are at a distance (< closest_pair_dis) from mid-point.
+(It is a proven that strip contains at most 6 points)
```
Comment (Erfaniaa) [inline divide_and_conquer/closest_pair_of_points.py]: No, it's wrong. Please review the theoretical part of this part more.

File: divide_and_conquer/closest_pair_of_points.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+Time complexity: O(n * (logn) ^ 2)
+"""
+
+
+import math 
+
+
+def euclidean_distance(point1, point2):
+    return math.sqrt(pow(point1[0] - point2[0], 2) + pow(point1[1] - point2[1], 2))
```
Comment (Erfaniaa) [inline divide_and_conquer/closest_pair_of_points.py]: It's OK but actually sqrt is unnecessary. For the optimization, you can use sqrt only once after all calculations have been done.

File: divide_and_conquer/closest_pair_of_points.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    closest_pair_dis = min(closest_in_left, closest_in_right)
+    
+    #points which are at a distance (< closest_pair_dis) from mid-point
+    cross_strip = []
+    for point in points:
+        if abs(point[0] - points[mid][0]) < closest_pair_dis:
+            cross_strip.append(point)
+
+    cross_strip = column_based_sort(cross_strip, 1)
+    closest_in_strip = dis_btw_closest_pair(cross_strip,
```
Comment (Erfaniaa) [inline divide_and_conquer/closest_pair_of_points.py]: Don't use brute force for the points inside the strip. You have to calculate this part in O(n).

File: divide_and_conquer/closest_pair_of_points.py (hunk 2 above)
Comment (Dharni0607) [reply divide_and_conquer/closest_pair_of_points.py]: ok.

**Comment by Erfaniaa** [comment]
The codes are now fine. I'm going to merge it but consider that we should reduce the closest pair of points time complexity to O(n log n), in future (now it's O(n * log n * log n)).

File: divide_and_conquer/closest_pair_of_points.py [hunk 4]
```diff
@@ -1,66 +1,113 @@
...
+        for j in range(i+1, points_counts):
+            current_dis = euclidean_distance_sqr(points[i], points[j])
             if current_dis < min_dis:
                 min_dis = current_dis
     return min_dis
 
 
-#divide and conquer approach
-def closest_pair_of_points(points, no_of_points):
+def dis_btw_closest_in_strip(points, points_counts, min_dis = float("inf")):
```
Comment (Erfaniaa) [inline divide_and_conquer/closest_pair_of_points.py]: It's not important but if you have a better name for this function, please update it. Long function names are also fine. Using "between" is better than "btw". Also you can use "brute force" word in its name.
These were only some suggestions. It's up to you.

File: divide_and_conquer/closest_pair_of_points.py [hunk 5]
```diff
@@ -1,66 +1,113 @@
...
-(It is a proven that strip contains at most 6 points)
-And brute force method is applied on strip to find closest points. 
+whose Xco-ords distance is less than closest_pair_dis
+from mid-point's Xco-ords.
+Closest pair distance is found in the strip of points. (closest_in_strip)
 
-Time complexity: O(n * (logn) ^ 2)
+min(closest_pair_dis, closest_in_strip) would be the final answer.
+ 
+Time complexity: O(n * (logn)^2)
```
Comment (Erfaniaa) [inline divide_and_conquer/closest_pair_of_points.py]: If you want you can put a reference to it. Are there any references in CLRS book?

**Comment by Dharni0607** [comment]
Ok.

**Comment by fty48jj19** [comment]
ok  obrigado saldaçoes  a todos ai  e um abraço

Em qui, 4 de jul de 2019 às 04:06, Erfan Alimohammadi <
notifications@github.com> escreveu:

**Pull request merged**